### PR TITLE
Impossible one ways Validator

### PR DESCRIPTION
**Pull request by Rub21**
from:  https://github.com/osmlab/osmlint/issues/85 
Result for US: https://gist.github.com/anonymous/4b76faeedf33c406c19e1026a4f23b5f
Time: 3:20 min for entire US(8xlarge)

@geohacker can you  take a look?

cc. @planemad @maning

### CODE REVIEW

**Comment by maning** [comment]
@Rub21 I did a quick review today, some observations below:

1. Let's ignore [paths](https://github.com/osmlab/osmlint/blob/imposibleoneways/validators/impossibleOneWays/map.js#L34), it does not make sense to have oneways on pedestrian routing.
2. I'm seeing a lot of false positives (see list below).  What I think it does is, it flags impossibleoneway if the oneway road  does not connect to another road.  We should ignore the starting point of the oneway and only look for road connection at the end point of the oneway direction.

* https://www.openstreetmap.org/way/342393160/history - goes to contsruction which is valid data
* https://www.openstreetmap.org/way/170771504/history - false positive
* https://www.openstreetmap.org/way/170771508/history - all roads going through this roundabout was flagged as impossible oneway
* https://www.openstreetmap.org/way/397090174/history - false positive
* https://www.openstreetmap.org/way/52646472/history - false positive
* https://www.openstreetmap.org/way/15390714/history - false positive

**Comment by Rub21** [comment]
@maning  New results forr USA : https://gist.github.com/anonymous/f739c1397428eb1b659e6e33962b426a

**Comment by maning** [comment]
Thanks @rub21, I reviewed all motorways and motorway_links today, below is the result:

| wayID | impossibleoneway | notes |  
|-----------------------------------------------------|------------------|---------------------------------------------------------------------------------|
| https://www.openstreetmap.org/way/290255689/history | true | but with access=no |  
| https://www.openstreetmap.org/way/118936937/history | true |  |  
| https://www.openstreetmap.org/way/365568444/history | true |  |  
| https://www.openstreetmap.org/way/365568445/history | true |  |  
| https://www.openstreetmap.org/way/18173499/history | true | missing segment connection to the roundabout |  
| https://www.openstreetmap.org/way/398643703/history | true | tunnel ending nowhere |  
| https://www.openstreetmap.org/way/44756199/history | true | small segment does not connect to the road |  
| https://www.openstreetmap.org/way/40394631/history | false | the error is the road south https://www.openstreetmap.org/way/364414704/history |  
| https://www.openstreetmap.org/way/364414702/history | false | same as above |  
| https://www.openstreetmap.org/way/364414704/history | true |  |  |
| https://www.openstreetmap.org/way/173416749/history | true | escape ramp shouldn't oneway tags (?) |  
| https://www.openstreetmap.org/way/90576393/history | true | missing service road to the parking lot |  
| https://www.openstreetmap.org/way/23465590/history | true | missing segment to connect to the tertiary road |  
| https://www.openstreetmap.org/way/365993236/history | true | island road - no start and end to other roads |  
| https://www.openstreetmap.org/way/48102651/history | true | end point does not connect to any road |  
| https://www.openstreetmap.org/way/19815539/history | true | motorway emergency road | should it have a oneway tag? |
| https://www.openstreetmap.org/way/407988615/history | true | tagged with access=no |  
| https://www.openstreetmap.org/way/177371683/history | true | end point does not connect to any road |  
| https://www.openstreetmap.org/way/14925228/history | true | tagged with access=no |  
| https://www.openstreetmap.org/way/407511489/history | true | tagged with access=no |  
| https://www.openstreetmap.org/way/407511490/history | true | but it was already tagged with access=no |  
| https://www.openstreetmap.org/way/114207087/history | true | tagged with access=emergency |  
| https://www.openstreetmap.org/way/85776410/history | true | tagged with access=emergency |  
| https://www.openstreetmap.org/way/113536663/history | true | tagged with access=emergency |  
| https://www.openstreetmap.org/way/14441108/history | true | end point does not connect to any road |  
| https://www.openstreetmap.org/way/389594882/history | true | end point does not connect to any road |  
| https://www.openstreetmap.org/way/404933423/history | true | tagged with disused=yes |  |
| https://www.openstreetmap.org/way/354416084/history | true | end point does not connect to any road |  
| https://www.openstreetmap.org/way/47283966/history | true | start road have wrong way direction |  
| https://www.openstreetmap.org/way/362806098/history | true | wrong way direction |  
| https://www.openstreetmap.org/way/57962193/history | true | end point does not connect to any road |  

Generally, I think the detection is :ok_hand:.  Some motorway_links were detected as impossibleoneway but its has access=no or access=emergency, I'm not sure if oneway=yes is a valid tag for emergency motorway ramps.  To filter out more falsepositives, can we remove highways that has access=no?

@rub21 @geohacker @planemad

**Comment by Rub21** [comment]
@maning  ,

I've added rules to avoid all highways which is `*_link`,  new result : https://gist.github.com/anonymous/2704598b50c018e37ef98acf195ca1b3

**Comment by maning** [comment]
@Rub21 output, looks good to me. over to @geohacker for code review.

**Comment by Rub21** [comment]
@geohacker , Is this ready to merge on master?

File: validators/impossibleOneWays/map.js [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+      }
+      var idWayL = id + 'L';
+      for (var j = 0; j < coordsWayL.length; j++) {
+        var positionL;
+        if (j === 0) {
+          positionL = 'first';
+        } else if (j === (coordsWayL.length - 1)) {
+          positionL = 'end';
+        } else {
+          positionL = 'midle';
```
Comment (geohacker) [inline validators/impossibleOneWays/map.js]: `midle` --> `middle` or `midpoint`?

**Comment by geohacker** [comment]
Hey @Rub21 sorry this took so long. But this looks good to me. I think we can merge.

**Pull request merged**